File: py-project/Solution18.py
# -*- coding:utf-8 -*-

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplication(self, pHead):
        # write code here
        if pHead == None or pHead.next == None:
            return pHead
        new_head = ListNode(-1)
        new_head.next = pHead
        pnode = pHead
        pre = new_head
        nex = None
        while pnode and pnode.next:
            nex = pnode.next
            if nex.val == pnode.val:
                while nex and pnode.val ==nex.val:
                    nex = nex.next
                pnode = nex
                pre.next = pnode
            else:
                pre = pnode
                pnode = nex

        return new_head.next

if __name__ == "__main__":
    ln1 = ListNode(1)
    ln2 = ListNode(1)
    ln3 = ListNode(1)
    ln4 = ListNode(2)
    # ln5 = ListNode(3)
    # ln6 = ListNode(4)
    # ln7 = ListNode(5)
    # ln8 = ListNode(5)
    ln1.next = ln2
    ln2.next = ln3
    ln3.next = ln4
    # ln4.next = ln5
    # ln5.next = ln6
    # ln6.next = ln7
    # ln7.next = ln8
    # print(ln2.next)
    pnode  = ln1
    while pnode:
        print(pnode.val)
        if pnode.next:
            pnode = pnode.next
        else:
            break
    print('\n')
    solution1 = Solution()
    a = solution1.deleteDuplication(ln1)
    pnode  = a
    while pnode:
        print(pnode.val)
        if pnode.next:
            pnode = pnode.next
        else:
            break
    # print(a)

File: c2-project/Solution18.cpp
//
// Created by 陈高杰 on 2019-11-15.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
#include <math.h>
using namespace std;


struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(nullptr) {
    }
};

class Solution {
public:
    ListNode* deleteDuplication(ListNode* pHead)
    {
        if(pHead == nullptr or pHead->next== nullptr){
            return pHead;
        }
        ListNode* new_head = new ListNode(-1);
        new_head->next = pHead;
        ListNode* pnode = pHead;
        ListNode* pre = new_head;
        ListNode* next = nullptr;
        while(pnode != nullptr && pnode->next != nullptr){
            next = pnode->next;
            if(next->val == pnode->val){
                while(next != nullptr && next->val == pnode->val) {
                    next = next->next;
                }
                pnode = next;
                pre->next = pnode;

            }
            else{
                pre = pnode;
                pnode = next;
            }
        }
        ListNode* output = new_head->next;
        delete new_head;
        return output;
    }
};

//int main()
//{
//    ListNode* l1 = new ListNode(1);
////    ListNode* l2 = new ListNode(2);
////    ListNode* l3 = new ListNode(2);
////    ListNode* l4 = new ListNode(3);
////    ListNode* l5 = new ListNode(1);
////    l1->next = l2;
////    l2->next = l3;
////    l3->next = l4;
////    l4->next = l5;
//    ListNode* pnode = l1;
//    while(pnode){
//        cout << pnode->val << endl;
//        pnode = pnode->next;
//    }
//    Solution solution;
//    ListNode* pnode2 = solution.deleteDuplication(l1);
//    while(pnode2){
//        cout << pnode2->val << endl;
//        pnode2 = pnode2->next;
//    }
//    return 0;
//}